AX: Add IPC to clear the isolated tree

hoffmanjoshua opened the pull request:
#### e230fe840ebe2ff2e07f282e95c0d93b42a2a2cf
<pre>
AX: Add IPC to clear the isolated tree
<a href="https://bugs.webkit.org/show_bug.cgi?id=295905">https://bugs.webkit.org/show_bug.cgi?id=295905</a>
<a href="https://rdar.apple.com/155804317">rdar://155804317</a>

Reviewed by Tyler Wilcock.

For debugging, adds a mechanism to destroy the isolated tree
for a given page. This means that the next time AX queries the
web page, a new tree is generated.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAccessibilityIsolatedTree):
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.h:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(WKPageAccessibilityClearIsolatedTree):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getAccessibilityWebProcessDebugInfo):
(WebKit::WebPageProxy::clearAccessibilityIsolatedTree):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearAccessibilityIsolatedTree):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: <a href="https://commits.webkit.org/297370@main">https://commits.webkit.org/297370@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9d6ac78eed2d12730c416ebd7f7010698a38e13

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/111517 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/31183 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/21648 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/117549 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/61789 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/113479 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/31864 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/39765 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/84739 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/35530 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/114464 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/25441 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/100378 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/65187 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/24788 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/18514 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/61373 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/94825 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/18579 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/120778 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/38566 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/28654 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/93658 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/38942 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/96646 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/93484 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23826 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/38601 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/16385 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/34603 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/38455 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/43932 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/38120 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/41453 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/39822 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/5cfe414bbfc032dbe046a6993e912ccd14b513d8. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47987

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/4cfeb720dfd0a798280eb816e9716bd2669da077. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47987

twilco commented on Source/WebCore/page/Page.cpp:
```diff
@@ -805,6 +805,16 @@ std::optional<AXTreeData> Page::accessibilityTreeData(IncludeDOMInfo includeDOMI
+void Page::clearAccessibilityIsolatedTree()
+{
+    if (CheckedPtr cache = axObjectCache()) {
+        if (auto identifier = this->identifier())
```
std::optional rather than auto?

twilco commented on Source/WebCore/page/Page.cpp:
```diff
@@ -805,6 +805,16 @@ std::optional<AXTreeData> Page::accessibilityTreeData(IncludeDOMInfo includeDOMI
+{
+    if (CheckedPtr cache = axObjectCache()) {
+        if (auto identifier = this->identifier())
+            AXIsolatedTree::removeTreeForPageID(*identifier);
```
Don't we need to re-build the tree too? Won't just clearing it cause all web content to be permanently inaccessible?

hoffmanjoshua replied:
Rebuilding eagerly isn't necessary, since VoiceOver (or any AT) will query the page calling `rootObjectForFrame` and then `getOrCreateIsolatedTree`, which will re-create it since there isn't an entry for that page.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/862339fd18bbd8fe2f48433632ed44ee904969ed. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47987

hoffmanjoshua commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1759,22 +1759,29 @@ void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+void WebPageProxy::clearAccessibilityIsolatedTree()
+{
+    protectedLegacyMainFrameProcess()->send(Messages::WebPage::ClearAccessibilityIsolatedTree(), webPageIDInMainFrameProcess());
```
Good call! I updated this to use the `forEachWebContentProcess` utility, which will let us send this to each process associated with the Page.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/15ba0683ada53d6de582e99cb858984e70325ddd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47987

twilco commented on Source/WebKit/WebProcess/WebPage/WebPage.cpp:
```diff
@@ -2749,6 +2749,13 @@ void WebPage::getAccessibilityWebProcessDebugInfo(CompletionHandler<void(WebCore
+#if ENABLE(ACCESSIBILITY_ISOLATED_TREE)
+void WebPage::clearAccessibilityIsolatedTree()
+{
+    protectedCorePage()->clearAccessibilityIsolatedTree();
```
Is a null-check necessary here?

cdumez replied:
`m_page` becomes null after `WebPage::close()` has been called. There are places where we do null check:
```
void WebPage::setNeedsDOMWindowResizeEvent()
{
    RefPtr page = m_page;
    if (!page)
        return;
```

The question is whether or not this code can run after `WebPage::close()` has been called.

hoffmanjoshua replied:
Hmm, I guess there's a possibility that this runs after `WebPage::close()` has been called, so I can add the null check to be extra safe.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/d9d6ac78eed2d12730c416ebd7f7010698a38e13. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/47987

cdumez commented on Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
```diff
@@ -1759,22 +1759,27 @@ void addResult(HashMap<WebCore::FrameIdentifier, Ref<WebCore::LegacyWebArchive>>
+    return @{
+        @"axIsEnabled": [NSNumber numberWithBool:result.isAccessibilityEnabled],
+        @"axIsThreadInitialized": [NSNumber numberWithBool:result.isAccessibilityThreadInitialized],
+        @"axLiveTree": result.liveTree.createNSString().get(),
```
This code looks right to me. `createNSString()` returns a RetainPtr so we would not leak. It is up to to the new Cocoa container to retain those strings.

hoffmanjoshua committed:
AX: Add IPC to clear the isolated tree
https://bugs.webkit.org/show_bug.cgi?id=295905
rdar://155804317

Reviewed by Tyler Wilcock.

For debugging, adds a mechanism to destroy the isolated tree
for a given page. This means that the next time AX queries the
web page, a new tree is generated.

* Source/WebCore/page/Page.cpp:
(WebCore::Page::clearAccessibilityIsolatedTree):
* Source/WebCore/page/Page.h:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.h:
* Source/WebKit/UIProcess/API/C/mac/WKPagePrivateMac.mm:
(WKPageAccessibilityClearIsolatedTree):
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::getAccessibilityWebProcessDebugInfo):
(WebKit::WebPageProxy::clearAccessibilityIsolatedTree):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::clearAccessibilityIsolatedTree):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:

Canonical link: https://commits.webkit.org/297370@main

webkit-commit-queue commented:
Committed 297370@main (e230fe840ebe): <https://commits.webkit.org/297370@main>

Reviewed commits have been landed. Closing PR #47987 and removing active labels.
